app/models/gsv/PanoHistoryTable.scala

```diff
@@ -0,0 +1,72 @@
+
+import models.utils.MyPostgresDriver.simple._
+import play.api.Play.current
+import java.sql.Timestamp
+import java.text.SimpleDateFormat
+import java.util.{Date, TimeZone}
+import scala.slick.lifted.ForeignKeyQuery
+import scala.util.Try
+
+case class PanoHistory(panoId: String, visitedTimestamp: Option[Timestamp], panoDate: String, locationCurrentPanoId: String)
+
+class PanoHistoryTable(tag: Tag) extends Table[PanoHistory](tag, "pano_history") {
+  def panoId: Column[String] = column[String]("pano_id", O.PrimaryKey)
+  def visitedTimestamp: Column[Option[Timestamp]] = column[Option[Timestamp]]("visited_timestamp", O.Nullable) 
+  def panoDate: Column[String] = column[String]("pano_date", O.NotNull)
+  def locationCurrentPanoId: Column[String] = column[String]("location_current_pano_id", O.NotNull)
+  def * = (panoId, visitedTimestamp, panoDate, locationCurrentPanoId) <> ((PanoHistory.apply _).tupled, PanoHistory.unapply)
+  def pano: ForeignKeyQuery[PanoHistoryTable, PanoHistory] = foreignKey("pano_history_location_current_pano_id_fkey", locationCurrentPanoId, TableQuery[PanoHistoryTable])(_.panoId)
+}
+
+object PanoHistoryTable {
+  val db = play.api.db.slick.DB
+  val panoHistoryTable = TableQuery[PanoHistoryTable]
+
+  /**
+    * Get all pano histories.
+    */
+  def selectAllPanoHistories(): List[PanoHistory] = db.withSession { implicit session =>
+    panoHistoryTable.list
+  }
+
+  /**
+    * Checking if a row exists with a specific panoId.
+    */
+  def getRowByPanoId(panoId: String): Option[PanoHistory] = db.withSession { implicit session =>
+    val query = for {
+      row <- panoHistoryTable if row.panoId === panoId
+    } yield row
+
+    query.firstOption
+  }
+ 
+  /**
+    * Save a pano history object to the PanoHistory table. If a pano history is submitted with a null visitedTimestamp, 
+    * check to see if a timestamp entry exists for that row already and keep existing timestamp.
+    */
+  def save(currentPanoId: String, visitedTimestamp: Option[Long], currentPanoDate: String, locationCurrentPanoId: String): Int = db.withSession { implicit session =>
+    var visitedTimestampOption: Option[Timestamp] = None
+    if (visitedTimestamp != null) {
+        val timestampValue: Long = visitedTimestamp match {
+            case Some(value) => value
+            case None => 0
+        }
+        val visitedTimestampObject: Timestamp = new Timestamp(timestampValue)
+        visitedTimestampOption = Some(visitedTimestampObject)
+    } else {
+      val rowOption = getRowByPanoId(currentPanoId)
+      rowOption match {
+        case Some(row) => {
+          if (row.visitedTimestamp != null) {
+            visitedTimestampOption = row.visitedTimestamp
+          }
+        }
+        case None => {}
+      }
+    }
+    val newPanoHistory: PanoHistory = PanoHistory(currentPanoId, visitedTimestampOption, currentPanoDate, locationCurrentPanoId)
+    val rowsAffected = panoHistoryTable.insertOrUpdate(newPanoHistory)
```

Comment:
What does the table look like if we visit an image and store the history, then later on a new image is added at that location and then we visit it and try to store the new history. Is the table left in a good state that's easy to understand? What will it look like? What _should_ it look like?

Reply:
Yeah so then all the previous panos that are already in the table (including the one we first visited) will have their location_current_pano_id field changed to be the id of that new pano that's at that location. That way, if you wanna query all the panos that share a location of some pano, you can just get every pano that shares it's location_current_pano_id.

Reply:
I'm pretty sure I tested for this and I got the right behaviour, will test it out a little more today though and make changes if it's not working!

Reply:
Yeah I just tested it out and it should work - there is an assumption here that every time a new pano gets added as the most recent pano of a location that whenever we fetch all the pano's for that location from the GSV API, it will give us each and every one of those panos (which we should all have in our database already). If this is true, then each time pano history data gets sent to the backend, then the id of the most recent pano at that location should replace all the existing pano's location_current_pano_id fields in the database.

Reply:
Ok gotcha. Instead of adding new rows, how about I go through every row with a location_current_pano_id equal to the one getting overridden and change it to be the new location_current_pano_id? Just thinking that might reduce a good amount of redundancy in storage albeit at the expense of some speed.

Reply:
Yeah I think that's fine!